pyramid_oereb/contrib/__init__.py

```diff
@@ -4,3 +4,54 @@
 This package is meant for contributions to the project, that are not part of the core libraries but may be
 useful in general.
 """
+import logging
+
+from functools import cmp_to_key
+
+
+log = logging.getLogger(__name__)
+
+
+def plr_sort_within_themes_by_type_code(extract):
+    """
+    This method sorts a plr list (while keeping the theme order) according to the TypeCode.
+
+    Args:
+        extract
+
+    Returns:
+        The sorted extract
+    """
+    real_estate = extract.real_estate
+
+    if log.isEnabledFor(logging.DEBUG):
+        for public_law_restriction in real_estate.public_law_restrictions:
+            log.debug("START plr_sort_within_themes_by_type_code: plr theme {}, type_code {}".
+                      format(public_law_restriction.theme.code, public_law_restriction.type_code))
+
+    def cmp(a, b):
+        if a.theme.code == b.theme.code:
+            # Only impact the order of elements which are in the same theme
+            value_a = a.type_code
+            value_b = b.type_code
+            if value_a and value_b:
+                if value_a < value_b:
+                    ret = -1
+                elif value_a == value_b:
+                    ret = 0
+                else:
+                    ret = 1
+            else:
+                ret = 0
+        else:
+            ret = 0
+        return ret
+
+    real_estate.public_law_restrictions = sorted(real_estate.public_law_restrictions, key=cmp_to_key(cmp))
```

Comment:
it is not the same as (use lamda function):
`real_estate.public_law_restrictions = sorted(real_estate.public_law_restrictions, key=lambda x: x.get('type_code', ''))`(whithout the whole `cmp` function ?)

In all cases, with this code and these names: `['Arboretum', 'Ärboretum', 'Belize']` you will get this result:
`['Arboretum', 'Belize', 'Ärboretum']`
Because Ä > Z. It is the wanted result ?

Reply:
I don't think we can use the simple solution in this case, because the goal is not to sort by type_code, but to keep the ordering of themes, and within the themes, sort by type_code. I couldn't find a way to do that without a custom comparator function - WDYT?

Reply:
Oh, I see. Ok. So I've not immediat better solution.
And what about umlaut ? it is correct like that ?

Reply:
There are different standards for that... for me, the way Python does it (consider Ä, Ö, ... as separate letters and therefore sort them after Z) is OK.